platform/linux-generic/include/odp/api/plat/packet_types.h

```diff
@@ -67,101 +67,130 @@ typedef enum {
-		uint64_t ipfrag:1;    /**< IP fragment */
-		uint64_t ipopt:1;     /**< IP optional headers */
+		uint64_t ipv4:1;      /* IPv4 */
+		uint64_t ipv6:1;      /* IPv6 */
+		uint64_t ip_bcast:1;  /* IP broadcast */
+		uint64_t ip_mcast:1;  /* IP multicast */
+		uint64_t ipfrag:1;    /* IP fragment */
+		uint64_t ipopt:1;     /* IP optional headers */
 
-		uint64_t ipsec:1;     /**< IPSec packet. Required by the
+		uint64_t ipsec:1;     /* IPSec packet. Required by the
 					   odp_packet_has_ipsec_set() func. */
-		uint64_t ipsec_ah:1;  /**< IPSec authentication header */
-		uint64_t ipsec_esp:1; /**< IPSec encapsulating security
+		uint64_t ipsec_ah:1;  /* IPSec authentication header */
+		uint64_t ipsec_esp:1; /* IPSec encapsulating security
 					   payload */
-		uint64_t udp:1;       /**< UDP */
-		uint64_t tcp:1;       /**< TCP */
-		uint64_t tcpopt:1;    /**< TCP options present */
-		uint64_t sctp:1;      /**< SCTP */
-		uint64_t icmp:1;      /**< ICMP */
-
-		uint64_t color:2;     /**< Packet color for traffic mgmt */
-		uint64_t nodrop:1;    /**< Drop eligibility status */
-
-		uint64_t l3_chksum_done:1; /**< L3 checksum validation done */
-		uint64_t l4_chksum_done:1; /**< L4 checksum validation done */
-		uint64_t ipsec_udp:1; /**< UDP-encapsulated IPsec packet */
+		uint64_t udp:1;       /* UDP */
+		uint64_t tcp:1;       /* TCP */
+		uint64_t tcpopt:1;    /* TCP options present */
+		uint64_t sctp:1;      /* SCTP */
+		uint64_t icmp:1;      /* ICMP */
+
+		uint64_t color:2;     /* Packet color for traffic mgmt */
+		uint64_t nodrop:1;    /* Drop eligibility status */
+
+		uint64_t l3_chksum_done:1; /* L3 checksum validation done */
+		uint64_t l4_chksum_done:1; /* L4 checksum validation done */
+		uint64_t ipsec_udp:1; /* UDP-encapsulated IPsec packet */
 	};
 
 } _odp_packet_input_flags_t;
 
+/*
+ * Additional packet flags
+ */
+typedef union {
+	/* All flags */
+	uint32_t all_flags;
+
+	struct {
+		uint32_t reserved1:     12;
+
+	/*
+	 * Packet output flags
+	 */
+		uint32_t l3_chksum_set:  1; /* L3 chksum bit is valid */
+		uint32_t l3_chksum:      1; /* L3 chksum override */
+		uint32_t l4_chksum_set:  1; /* L4 chksum bit is valid */
+		uint32_t l4_chksum:      1; /* L4 chksum override  */
+		uint32_t shaper_len_adj: 8; /* Adjustment for traffic mgr */
+
+	/*
+	 * Error flags
+	 */
+		uint32_t snap_len_err:   1; /* Snap length error */
+		uint32_t ip_err:         1; /* IP error */
+		uint32_t l3_chksum_err:  1; /* L3 checksum error */
+		uint32_t tcp_err:        1; /* TCP error */
+		uint32_t udp_err:        1; /* UDP error */
+		uint32_t l4_chksum_err:  1; /* L4 checksum error */
+		uint32_t ipsec_err:      1; /* IPsec error */
+		uint32_t crypto_err:     1; /* Crypto packet operation error */
+	};
+
+	/* Flag groups */
+	struct {
+		uint32_t reserved2:     12;
+		uint32_t other:         12; /* All other flags */
+		uint32_t error:          8; /* All error flags */
+	} all;
+
+} _odp_packet_flags_t;
```

Comment:
Hmm. What about the following code:
```c
typedef struct {
  union {
   struct {
    // ...
   };
   uint32_t all_error : 8;
  } error;
  union {
    sruct {
    // ...
    };
    uint32_t all_other;
  } other;
} _odp_packet_flags_t;
```

I have not tried, but I think this should work and still provide cleaner separation between error non-error flags.

Reply:
To much nesting of structs and unions tends to make things harder to follow. Two deep is clearer than a three deep definition. I think I prefer the original.

Reply:
Dmitry, those two unions (error and other) would result two 32 bit words, which would defeat the purpose of combining flags into one word. If the top level would be union, it would be single word but now each group would need to be padded:
```
union foo {
	union {
		struct {
			uint32_t bit1:  1;
			uint32_t bit2:  1;
			uint32_t res1: 30;
		};

		struct {
			uint32_t all:   2;
			uint32_t res2: 30;
		};

	} bar;

	union {
		struct {
			uint32_t res1:  2;
			uint32_t bit1:  1;
			uint32_t bit2:  1;
			uint32_t res2: 28;
		};

		struct {
			uint32_t res3:  2;
			uint32_t all:   2;
			uint32_t res4: 28;
		};

	} bar2;
};

```

Since this is implementation internal, we can relay on developer remembering e.g. that xxx_err is an error flag, use a clean/flat struct. Also flag naming can be updated any time (since it's not API).

Reply:
@psavol Sorry, I though that C compiler would be able to merge those two unions into consecutive bit space. I experimented with it a bit and it seems you are correct. I can not find an easy way to force gcc into using following bits. Only using `uint8_t` helps.